File: thisboot/src/main/java/com/brageast/blog/thisboot/service/impl/ArticleServiceImpl.java
package com.brageast.blog.thisboot.service.impl;

import com.brageast.blog.thisboot.entity.Article;
import com.brageast.blog.thisboot.service.ArticleService;
import com.brageast.blog.thisboot.util.EntityUtil;
import com.mongodb.client.result.DeleteResult;
import com.mongodb.client.result.UpdateResult;
import lombok.extern.slf4j.Slf4j;
import org.bson.types.ObjectId;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.mongodb.core.ReactiveMongoTemplate;
import org.springframework.data.mongodb.core.query.Criteria;
import org.springframework.data.mongodb.core.query.Query;
import org.springframework.data.mongodb.core.query.Update;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.util.Date;

@Slf4j
@Service
public class ArticleServiceImpl implements ArticleService {

    @Autowired
    private ReactiveMongoTemplate reactiveMongoTemplate;

    /**
     *
     * @param page 默认减一, 使页面成 1 开始
     * @param size 每页的大小
     * @return
     */
    @Override
    public Flux<Article> limitShow(int page, int size) {
        return reactiveMongoTemplate.find(new Query().with(PageRequest.of(page - 1, size)), Article.class);
    }

    @Override
    public Mono<Article> insert(Mono<Article> article) {
        return reactiveMongoTemplate.insert(article);
    }

    @Override
    public Mono<Article> insert(Article article) {
        return reactiveMongoTemplate.insert(article);
    }

    @Override
    public Flux<Article> show() {
        return reactiveMongoTemplate.findAll(Article.class);
    }

    @Override
    public Mono<DeleteResult> delete(ObjectId id) {
        return reactiveMongoTemplate.remove(doId(id.toString()), Article.class);
    }

    @Override
    public Mono<UpdateResult> update(Article article) {
        Update update = EntityUtil.addUpdate(
                article, "title", "label", "classify",
                "describe", "author", "content", "chick", "replys"
        );
        update.set("upDate", new Date());

        return reactiveMongoTemplate.updateFirst(doId(article.getArticleId().toString()), update, Article.class);
    }

    @Override
    public Mono<Article> findById(ObjectId articleId) {
        return reactiveMongoTemplate.findOne(doId(articleId.toString()), Article.class);
    }

    private Query doId(String id ) {
        return Query.query(
                new Criteria().and("articleId").is(id)
        );
    }
}

File: thisboot/src/main/kotlin/com/brageast/blog/thisboot/service/impl/ArticleServiceImpl.kt
package com.brageast.blog.thisboot.service.impl

import com.brageast.blog.thisboot.entity.Article
import com.brageast.blog.thisboot.repository.ArticleRepository
import com.brageast.blog.thisboot.service.ArticleService
import org.bson.types.ObjectId
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.domain.PageRequest
import org.springframework.data.mongodb.core.ReactiveMongoTemplate
import org.springframework.stereotype.Service
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono


@Service
class ArticleServiceImpl (
    val articleRepository: ArticleRepository,
    val reactiveMongoTemplate: ReactiveMongoTemplate
) : ArticleService {

    override fun findAll(): Flux<Article> = articleRepository.findAll()
    /**
     * 用reactiveMongoTemplate 实现为 reactiveMongoTemplate.
     *                              find(Query()
     *                              .with(PageRequest.of(page - 1, size)), Article::class.java)
     *
     * @param page 默认减一, 使页面成 1 开始
     * @param size 每页的大小
     * @return
     */
    override fun findWithPagination(page: Int, size: Int): Flux<Article> {
        return if(size <= 0 || page < 0) Flux.empty() else articleRepository.findAllBy(PageRequest.of(page - 1, size))
    }

    override fun insert(article: Article): Mono<Article> = reactiveMongoTemplate.insert(article)

    override fun updateOrInsert(article: Article): Mono<Article> = articleRepository.save(article)

    override fun findById(id: ObjectId): Mono<Article> = articleRepository.findById(id)

    override fun deleteById(id: ObjectId): Mono<Void> = articleRepository.deleteById(id)

    override fun findById(id: String): Mono<Article> {
        val objectId = ObjectId(id)
        return articleRepository.findById(objectId)
    }

    /*override fun update(article: Article): Mono<UpdateResult> = article.run {
        val parse = Document.parse(this.toJSON())

        val update = Update.fromDocument(
                parse, "articleId", "update", "startDate", "chick"
        ).set("upDate", Date())

        val query = articleId?.toQueryById("articleId") ?: return@run Mono.just(UpdateResult.unacknowledged())

        return reactiveMongoTemplate.updateFirst(query, update, Article::class.java)
    }*/

}